dockerclient.go

```diff
@@ -212,6 +212,50 @@ func (client *DockerClient) ContainerChanges(id string) ([]*ContainerChanges, er
 	return changes, nil
 }
 
+func (client *DockerClient) readJSONStream(stream io.ReadCloser, decode func(*json.Decoder) decodingResult) (<-chan decodingResult, chan<- struct{}) {
+	resultChan := make(chan decodingResult)
+	closeChan := make(chan struct{})
+	go func() {
+		defer close(resultChan)
+
+		internalResultsChan := make(chan decodingResult)
+		defer close(internalResultsChan)
+
+		stillListening := make(chan struct{})
+		defer close(stillListening)
+
+		go func() {
```

Comment:
What happens if you get rid of the nested goroutine and do that in the main goroutine?

After `decode` returns you can check the closeCh:
```go
decodeResult := decode(decoder)
select {
    case <-closeChan:
        return
    default:
}
```

In both cases anyway, when you `closeCh`, you still have a lingering goroutine until `decode` returns